asaplib/cluster/ml_cluster_fit.py

```diff
@@ -100,115 +102,344 @@ def pack(self):
+
+        # The number of neighbours to consider is dictated by the percent parameter provided (default is 2% of the
+        # total data points.
+        n_neigh = int(self.percent/100.*Nele)
+
+        # Compute with 4 times the recommended percentage of neighbours in order that we may re-estimate the percent
+        # value supplied (rper) based on the data.
+        n_neigh_comp = int(4.*self.percent/100.*Nele)
+        neigh = NearestNeighbors(n_neighbors=n_neigh_comp).fit(data)
+        self.distances, self.indices = neigh.kneighbors(data)
+
+        # The cutoff distance dc is the average distance (averaged over all data points) of the "0.02*Nele"th neighbour
+        dc = np.average(self.distances[:, n_neigh])
+
+        # To store the local densities of each data point
+        dens = np.zeros(data.shape[0])
+        tt = 1.0
+        factor = 1.0
+
+        # We update dc based on a re-estimated percentage rper. Stop when rper is close to self.percent
+        while tt > 0.05:
+            dc = dc*factor
+            for i in range(Nele):
+                a = self.distances[i, :]
+                dens[i] = len(a[(a <= dc)])
+            rper = 100.*np.average(dens)/Nele
+            tt = rper/self.percent - 1.
+            if tt > 0.:
+                factor = factor/1.001
+            else:
+                factor = factor*1.001
+            tt = abs(tt)
+        self.dc = dc
+
+    def get_decision_graph(self, data):
+
+        Nele = data.shape[0]
+        self.dens = np.zeros(Nele)
+
+        if self.dc == None:
+            self.get_dc(data)
+
+        else:  # check compatibility between the provided cutoff distance and the provided distance matrix
+            n_neigh_comp = int(10.*self.percent/100.*Nele)
+            neigh = NearestNeighbors(n_neighbors=n_neigh_comp).fit(data)
+            self.distances, self.indices = neigh.kneighbors(data)
+            if self.dc > np.min(self.distances[:, n_neigh_comp - 1]):
+                print("dc too big for being included within the", 10.*self.percent, "%  of data, consider using a "
+                                                                                    "small dc or augment the percent "
+                                                                                    "parameter")
+
+        for i in range(Nele):
+            a = self.distances[i, :]
+            self.dens[i] = len(a[(a <= self.dc)])
+
+        if self.dens_type == 'exp':  # Density is no longer number but exponential
+            for i in range(Nele):
+                a = self.distances[i, :]/self.dc  # distances expressed as a fraction of the cutoff distance
+                self.dens[i] = np.sum(np.exp(-a**2))
+
+        # The densities computed by this class or not log densities. Log densities are returned by the KDE class.
+
+        # self.dens_cut = 0.2 * np.mean(np.log(self.dens)) + 0.8 * np.min(np.log(self.dens))
+        self.dens_cut = 0.2 * np.mean(self.dens) + 0.8 * np.min(self.dens)
```

Comment:
There's a difference here between the `LAIO_DB` class and your old class in that your old class takes in the log densities returned from the `KDE` class whereas the LAIO_DB class computes the densities itself in real (non-log) space. I'm not sure yet what implications this has for the heuristic formula for `self.dens_cut`